### ISSUE TITLE
Benchmarking Katana Performance

### ISSUE THREAD

**Issue by neotheprogramist** [neotheprogramist/dojo#7]
## Issue Description:
The objective of this issue is to benchmark the transaction per second (TPS) for calling the `move` operation with `katana`.

### Specifics:
- **Goal**: To determine the maximum throughput of `katana` in a 3-second block.
- **Method**: Populate the block with transactions from multiple accounts, all trying to call the `move` operation.
- **Constraint**: There should be no gas limit to truly test the limits of what is achievable.

### Questions to Address:
- What is the maximum number of transactions `katana` can handle in a 3-second block?
- How does the system behave under high transaction load?

## Expected Outcome:
This benchmarking will help us understand the capabilities and limitations of `katana`, especially in terms of handling high transaction volumes within a short time frame.

**Comment by matzayonc** [comment]
Using prefunded prefunded accounts and preparing transactions to spam them as fast as possible I managed to process 500 transactions in about 300ms.

With a greater number of transaction I experienced races and nonces being out of order. Batching should be enough to counteract that  as one thread will send all transactions from a single account, assuming no network races.

**Comment by glihm** [comment]
> Using prefunded accounts and preparing transactions to spam them as fast as possible I managed to process 500 transactions in about 300ms.
> 
> With a greater number of transaction I experienced races and nonces being out of order. Batching should be enough to counteract that as one thread will send all transactions from a single account, assuming no network races.

Sounds exciting! May I ask you some details on that?

1. Are these transactions with 1 call only?
2. Which parameters did you used for this benchmark? Fee disabled + transaction signature validation?

@matzayonc do you think there is a way to have the benchmark also expressed as Step per Seconds? Or maybe it's a totally different one? Just wanted to mention it in case it gives you some additional ideas for testing this specific case.

As multicall biases a bit the TPS metric in some situations (I assume in the bench you're doing it's not an issue), it could be great to have the Step per Seconds unit too.

Also, testing the max step per second may be a totally different setup / testing strategy.

**Comment by matzayonc** [comment]
> Sounds exciting! May I ask you some details on that?

I agree! Yes, for this benchmark I will let test the limits of TPS. Adding one with many and/or different ones should be trivial after the first one is finished. 
I use all default parametrs and have in mind to compare them when it is fully operational.

Steps per Seconds also sounds like a good idea could you exapand a bit on your idea?

**Comment by glihm** [comment]
> Yes, for this benchmark I will let test the limits of TPS
Sounds great, which kind of transaction do you use? An ERC20 transfer?

> I use all default parametrs and have in mind to compare them when it is fully operational.
Yes I agree, having results with different parameters tuned would be great (especially for `--disable-fee` and `--disable-validate`).

> Steps per Seconds also sounds like a good idea could you exapand a bit on your idea?
Due to the multicall feature + the very variable amount of work required by the network for a transaction can make the TPS not that accurate.
TPS is great for a very known transaction type (a transfer, a swap), but when it comes to network capabilities, having the max step per seconds is totally independent of the transaction type.

So, to test the max step per seconds, a naive approach may be to have only one (or few) transaction sent, which triggers a huge amount of work on the network (that we know/estimate in advance), and evaluate how much time Katana takes to complete it. This should give us the maximum steps Katana can process in a unit of time (let's say second).

Ideally we should find a way to have this huge work being something diversified / with some randomness to avoid possible cache / CPU optimizations.

Does it make sense to you too?

**Comment by matzayonc** [comment]
Apologies for the late response.

> Sounds great, which kind of transaction do you use? An ERC20 transfer?

For the main one the `move` operation from the `sozo init` contract

> Does it make sense to you too?

Sure, more focus on the VM performance, with minimizing the network overhead. I will be sure to include a few cases of this type.

I believe building a merkle tree or prime factorization would be a reasonable choice, do you agree?

**Comment by glihm** [comment]
> I believe building a merkle tree or prime factorization would be a reasonable choice, do you agree?

Let's do it then, both are good in this situation and you're surely better positioned than me to decide, it's at your discretion. :+1:
